drivers/soundwire/intel.c

```diff
@@ -1172,40 +1188,117 @@ static int intel_probe(struct platform_device *pdev)
 
 err_interrupt:
 	sdw_cdns_enable_interrupt(&sdw->cdns, false);
-	free_irq(sdw->link_res->irq, sdw);
 err_init:
-	sdw_bus_master_delete(&sdw->cdns.bus);
 	return ret;
 }
 
-static int intel_remove(struct platform_device *pdev)
+static void intel_master_remove(struct sdw_intel_master_dev *master_dev)
 {
-	struct sdw_intel *sdw;
+	struct device *dev = sdw_intel_master_to_dev(master_dev);
+	struct sdw_intel *sdw = master_dev->pdata;
+	struct sdw_bus *bus = &sdw->cdns.bus;
 
-	sdw = platform_get_drvdata(pdev);
-
-	if (!sdw->cdns.bus.prop.hw_disabled) {
+	if (!bus->prop.hw_disabled) {
 		intel_debugfs_exit(sdw);
 		sdw_cdns_enable_interrupt(&sdw->cdns, false);
 		free_irq(sdw->link_res->irq, sdw);
-		snd_soc_unregister_component(sdw->cdns.dev);
+		snd_soc_unregister_component(dev);
 	}
-	sdw_bus_master_delete(&sdw->cdns.bus);
+	sdw_bus_master_delete(bus);
+}
+
+static const struct sdw_intel_link_ops intel_master_link_ops = {
+	.startup = intel_master_startup,
+};
+
+#if IS_ENABLED(CONFIG_VIRTUAL_BUS)
+static int intel_master_probe_vdev(struct virtbus_device *vdev)
+{
+
+	struct sdw_intel_master_dev *master_dev =
+		to_sdw_intel_master_dev(vdev);
+
+	return intel_master_probe(master_dev);
+}
+
+static int intel_master_remove_vdev(struct virtbus_device *vdev)
+{
+
+	struct sdw_intel_master_dev *master_dev =
+		to_sdw_intel_master_dev(vdev);
+
+	intel_master_remove(master_dev);
 
 	return 0;
 }
 
-static struct platform_driver sdw_intel_drv = {
-	.probe = intel_probe,
-	.remove = intel_remove,
-	.driver = {
-		.name = "int-sdw",
+static void intel_master_shutdown_vdev(struct virtbus_device *vdev)
+{
+	/* FIXME: what should we do here */
```

Comment:
@plbossart we should do the same as the release function here ie free the master device data

Reply:
you mean call the release explicitly? or unregister the device?